### DISCUSSION TITLE
Factories with deep relations?

### DISCUSSION THREAD

**Discussion by genu** [adonisjs/core#2757] [Help]
Is there a way to Create factories with deep relations?

For example

```
A.with('B', 1).with('C', 1).create()
D.with('C', 1)
```

Is there a way to relate the `C`'s in the `A` factory and `D` factory if the`C` model looks like:

|a_id | d_id |
| -  |  - |
| ||

With the above factories, it actually produces:
|a_id | d_id |
-|-|
| 1  |  - |
| -|2|

Is there away to write the factory in such away that it creates Both `A` and `D` with refer to a single `C`?

**Comment by thetutlage** [comment]
Can you share a real example. It is hard to form thoughts with abstract names like `A`, `B`, `C`

**Reply by genu** [reply]
Sure, 

consider these models:

Company:
```ts
export default class Company extends BaseModel {
  @column({ isPrimary: true })
  public id: number

  @column()
  public name: string

  @hasMany(() => Invoice)
  public invoices: HasMany<typeof Invoice>
}
```
Client:
```ts
export default class Client extends BaseModel {
  @column({ isPrimary: true })
  public id: number

  @column()
  public userId: number

  @hasMany(() => Invoice)
  public invoices: HasMany<typeof Invoice>
}
```

Invoice:
```ts
export default class Invoice extends BaseModel {
  @column({ isPrimary: true })
  public id: number

  @column()
  public clientId: number

  @column()
  public companyId: number

  @belongsTo(() => Company)
  public company: BelongsTo<typeof Company>

  @belongsTo(() => Client)
  public client: BelongsTo<typeof Client>
}
```

Could I write a `Client` and `Company` factory in such a way that a shared `Invoice` is produced?

Using the factory in this way, technically, is incorrect because an `Invoice` must have a company and a client:

```ts
ClientFactory.with('invoice', 1); // Creates a client with an invoice, but not associated with a company
CompanyFactory.with('invoice', 1); // Creates a company with an invoice, but not associated with a client
```

**Reply by thetutlage** [reply]
Since Invoice is the common with both Client and Company. How about approaching it as follows.

```ts
await Invoice.with('company', 1).with('client', 1).create()
```

**Reply by genu** [reply]
That's how I figured it would need to work.

What about if there is a N:M table that relates `Client` and `Company`? 

Starting from the `Invoice` factory and working outward, doesn't actually create that N:M relationship that is needed.

Unless its done in extra steps after the initial factory.

Correct?
